### ISSUE TITLE
stylix: support slideshows and animated wallpapers

### ISSUE THREAD

**Issue by Lord-Valen** [danth/stylix#63]
My understanding is that stylix currently uses `stylix.image` for palette generation and wallpapers, and is mandatory. If one wishes to use `stylix.autoEnable`, they might be surprised to see that stylix is also setting wallpaper options. For users who wish to use a rotating set of wallpapers, this can be an annoyance as they would have to manually disable such modules. To accomodate users with sets of wallpapers, stylix should continue to use `stylix.image` but accepting a list, or expose a new option (`stylix.wallpapers`? `stylix.images`? Seems to complicate things for little benefit) which is a list of images. In order to accomplish the first solution (which I prefer), I would sugget taking the first or last element of the list when only one image is required.

**Comment by dwarfmaster** [comment]
A bit related to #54 . The point is that people want other things than just an image as wallpaper, so we need to support that. Maybe we could have an options `stylix.wallpaper` that is pretty free-form, and set `stylix.image` from it if it is set. I'll experiment with this idea and open a draft PR.

**Comment by danth** [comment]
I agree, there should be a single option which takes one of

- An image
- A list of images
- An animation

Theoretically the palette generator could support all of these directly, as it's just selecting pixels at random and then improving on that selection. That would however be quite resource intensive, especially for video wallpapers, as the entire file needs to be loaded into memory to do the selection.

I think we should stick with the idea of picking a single image/frame to use in the generator for now.

How should we handle backwards compatibility with the creation of a new option? `stylix.wallpaper` cannot default to `stylix.image`, because the default image will depend on the wallpaper, which creates an infinite loop if neither of the options are set.

**Comment by dwarfmaster** [comment]
One limitation with this approach is that in the case of the list, we may want more features (for example how long to display each wallpaper, how to handle transitions...). I don't really have any good idea on how to handle it.

**Comment by danth** [comment]
Is it possible to use a submodule type within an enum? If so, we could replace the list with a submodule having those extra options.

**Comment by dwarfmaster** [comment]
We can as long as there is only one such submodule (otherwise it gets complicated pretty fast I think), so it would work but if we want to support other kinds of wallpaper in the future it may get complicated, but I'm not sure how probable it is.

**Comment by danth** [comment]
On further thought, we could represent all wallpapers as an attrset:

```nix
{
  type = "image";
  file = ./image.png;
}
```

```nix
{
  type = "animation";
  file = ./video.mp4;
}
```

```nix
{
  type = "slideshow";
  delay = 5;
  files = [ ./image1.png ./image2.png ];
}
```

[It's possible to define custom types with a validator function](https://github.com/NixOS/nixpkgs/blob/ca39994de4a91b32c586191ad8f6990594926218/lib/types.nix#L108-L177), which in this case would reject anything without a `type` attribute or not matching the specification for that type.

We can then add constructors to simplify and provide defaults:

```nix
mkImage ./image.png
```

```nix
mkAnimation ./video.mp4
```

```nix
mkSlideshow {
  files = [ ./image1.png ./image2.png ];
}
```

This also avoids the need to parse file extensions, as the type of wallpaper is known.

What do you think?

Also, should there be a separate type for animated images (GIF, AVIF) vs video (MP4, AVI)? As these might need a different implementation.

**Comment by SomeGuyNamedMy** [comment]
it may be best to separate animated images and videos as mpvpaper is much less efficient then its cousins that only support animated images, at least on wayland, I dont know too much about the x11 side of things, i would recommend swww for the implementation of the slideshow and animated image types if we plan to go forward with this solution.

**Comment by SomeGuyNamedMy** [comment]
is this the plan for implementing this? because i think i might try my hand at it soon.

**Comment by danth** [comment]
Custom types are the best solution I can come up with. Let's go ahead with that.

**Comment by SomeGuyNamedMy** [comment]
whats everyone's opinion on maybe moving the color scheme to the same attribute set as the image? then rather then setting the color scheme directly you use a constructor to set the base16 scheme like nord or any other scheme.

**Comment by dwarfmaster** [comment]
I think @danth's idea is great but I'd like to mention it is pretty easy right now to use official themes: see how I do it [here](https://github.com/dwarfmaster/home-nix/blob/master/profiles/nixos/interface/theme/default.nix#L60).

One problem with using custom types is that it makes the option less self-documenting, so it should be avoid as much as possible. I actually don't know of any module where it's used, but it would be great to find some previous uses to see how they did it.

**Comment by SomeGuyNamedMy** [comment]
you guys are probably more knowledgable in this subject then i am so i dont really have much of an opinion on the matter

**Comment by SomeGuyNamedMy** [comment]
im personally in favor of moving everything over to the constructor but @danth what do you think?

**Comment by danth** [comment]
I'm also in favor of the custom types, but it would be nice if we had a way to generate documentation for them

**Comment by SomeGuyNamedMy** [comment]
@danth it looks like the gnome module relies on the polarity option, perhaps it would be a good idea to retain the polarity option but not have it influence the generated colorscheme? because there are other modules that i can see said option being useful for like qutebrowser.

**Comment by danth** [comment]
(As mentioned on Matrix) I think we should calculate polarity based on the background colour when it's needed for a module, as this will support handmade schemes too. So if the background colour is darker than a threshold, set the theme to dark, otherwise set the theme to light. Then the polarity option will only be used as a guide for the generated scheme.

**Comment by trueNAHO** [comment]
*Dumping additional slideshow feature suggestions.*

1. Randomly select an item from collection ([`mkSlideshow.files`](https://github.com/danth/stylix/issues/63#issuecomment-1486460669)).
1. Use a blurred image copy for the borders instead of black borders, if the aspect ration of the item does not match the aspect ratio of the drawing canvas.
1. Add additional options for randomly selecting items (feature suggestion `1.`): select only items that match the specified [`stylix.base16Scheme`](https://danth.github.io/stylix/configuration.html#handmade-schemes) scheme in order to remain in the same color scheme.

On non-nix-based systems, I currently use the following shell script to implement the first two feature suggestions: https://github.com/trueNAHO/dotfiles/blob/767b4ba1cc84ee2c980956cab72c708a958087f9/home/dot_local/bin/executable_set_wallpaper

**Comment by cramt** [comment]
i've implemented video background in my own setup (sway) i specifically had to solve the issue of differently sized monitors, so if anyone wants to steal that for a proper stylix solution please be my guest https://github.com/cramt/nixconf/blob/main/homeManagerModules/features/sway/default.nix#L6-L29

**Comment by beh-10257** [comment]
Is stylix.image still mandatory ???
 can't I just specify my own 16 colors and enjoy the theming part of stylix

I still didn't install stylix I'm asking before switching my current setup

**Comment by trueNAHO** [comment]
> Is stylix.image still mandatory ??? can't I just specify my own 16 colors and enjoy the theming part of stylix

Yes, `stylix.image` is still mandatory. In the future, it will become optional:

> - [ ] **(3)** Optionalize wallpaper functionality
>    - https://github.com/danth/stylix/issues/200
>    - https://github.com/danth/stylix/issues/442
>    - https://github.com/danth/stylix/pull/717
>
> -- https://github.com/danth/stylix/issues/534

**Comment by danth** [comment]
@beh-10257 Update - it's now optional and can be set to `null`

**Comment by sadorowo** [comment]
Is somebody working on this? I've recently moved from Pywal and the only thing that I am missing is the ability to change wallpaper every couple of hours. This would be very powerful feature, especially when combined with animated backgrounds or something similar.

**Comment by trueNAHO** [comment]
> Is anybody working on this?

Not yet. But it is on the roadmap:

> - [ ] **(6)** Implement slideshow with post-processing capabilities
>
> -- https://github.com/danth/stylix/issues/534

> I've recently moved from Pywal and the only thing that I am missing is the ability to change wallpaper every couple of hours. This would be very powerful feature, especially when combined with animated backgrounds or something similar.

Indeed. I assume that this functionality is the most desired one alongside:

> - [ ] **(4)** Support base24 scheme and Vim highlight groups to resolve styling inconsistencies
>
> -- https://github.com/danth/stylix/issues/534

IMHO, it is long-term better for this project to clean up the code base by resolving the previous roadmap tasks before implementing this feature, to avoid reaching an unmanageable amount of complexity.
